Ds feat/oracle mvp

0xDiscotech opened the pull request:
Oracle MVP solution

0xOneTony commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,301 @@
+    }
+
+    // Update the default wrapper for the pair
+    _pairInfo.defaultWrapperName = _wrapperName;
```
It's good practise to emit events when you change the storage value of a variable

0xOneTony commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,301 @@
+      return 0;
+    }
+    // Get the wrapper name for the pair
+    bytes32 _tokensPairHash = keccak256(abi.encodePacked(_tokenIn, _tokenOut));
```
Let's say you have a pair `ETH-USDC`, what happens if I give `tokenIn=USDC` and `tokenOut=ETH` or the reverse?

0xDiscotech replied:
Fixed now. I  check also for the hash in the inverse pair in case the pair id is zero 👍

0xOneTony commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,301 @@
+  ///@inheritdoc IOracle
+  uint256 public constant ONE_ETH = 1 ether;
+  ///@inheritdoc IOracle
+  bytes10 public constant INCH = bytes10('1inch');
```
Is there a specific reason `1inch` is in lowercase while the rest oracle names are in uppercase?

0xDiscotech replied:
It is a mistake

0xOneTony commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,301 @@
+
+    // Get the pair default wrapper
+    if (_pairInfo.defaultWrapperName != bytes10('')) {
+      // _wrapperName = _pairInfo.defaultWrapperName;
```
remove comment

0xOneTony commented on solidity/interfaces/IOracle.sol:
```diff
@@ -0,0 +1,205 @@
+ * @author 0xDiscotech
+ * @notice This contract manages different oracle wrappers for different pairs
+ */
+interface IOracle {
```
Since you have natspec comments here, you can use `@inheritdoc` in `Oracle.sol`'s functions

0xOneTony commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,301 @@
+  }
+
+  // Function that returns the wrapper name for the given pair
+  function _getWrapperName(PairInfo memory _pairInfo) internal view returns (bytes10 _wrapperName) {
```
Missing natspec for internal functions

0xOneTony commented on solidity/test/e2e/Oracle.t.sol:
```diff
@@ -0,0 +1,139 @@
+import {IinchOracle} from 'interfaces/IinchOracle.sol';
+import {OracleHelper} from '../../utils/OracleHelper.sol';
+
+contract E2EOracle is CommonE2EBase {
```
🔥

0xOneTony reviewed (commented):
🔥 I like your tests, left a few comments and questions

0xDiscotech commented:
> 🔥 I like your tests, left a few comments and questions

Thanks Tony, very useful comments

0xDiscotech commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,301 @@
+    address _tokenTwo,
+    bool _pairDefautlWrapper
+  ) external wrapperIsValid(_wrapperName) onlyOwner {
+    if (_wrapper == address(0) || _tokenOne == address(0)) {
```
Because it can be `address(0)` in case you only want to add for one token

0xDiscotech commented:
Pushed fixes based on @0xOneTony suggestions and comments

0xng commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,344 @@
+  ///@inheritdoc IOracle
+  uint256 public constant ONE_ETH = 1 ether;
+  ///@inheritdoc IOracle
+  bytes10 public constant INCH = bytes10('1INCH');
```
Just a small note here, if I recall correctly, constants and immutables are padded to be 32 bytes in the bytecode, so if you used bytes10 for any other reason than saving a slot with `globalDefaultWrapper` know that you can probably have used bytes32 if you wanted

0xDiscotech replied:
Good to know that. Yes, I used `bytes10` for saving a slot

0xng commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,344 @@
+    }
+
+    // Check if the wrapper already exists
+    PairInfo memory _pairInfo = getPairInfo(_tokenOne, _tokenTwo);
```
You can probably combine these two lines:
` uint104 _pairId = getPairInfo(_tokenOne, _tokenTwo).pairId;`

drgorillamd commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,344 @@
+import {OracleHelper} from './utils/OracleHelper.sol';
+import {IOracle} from '../interfaces/IOracle.sol';
+
+contract Oracle is IOracle, Ownable {
```
inheritance ordering (cf style guide or OZ https://github.com/OpenZeppelin/openzeppelin-contracts/blob/4c713f8cea2cbd002509c8ea0d8e0a996627f069/contracts/governance/Governor.sol#L28)

drgorillamd commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,344 @@
+
+contract Oracle is IOracle, Ownable {
+  ///@inheritdoc IOracle
+  uint256 public constant ONE_ETH = 1 ether;
```
really needed':)?

0xDiscotech replied:
Absolutely not, I will remove it

drgorillamd commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,344 @@
+    PairInfo storage _pairInfo = pairInfo[_tokensPairHash];
+    uint104 _pairId = _pairInfo.pairId;
+
+    // Check if the pair exists, but the order of the tokens is the inverse
```
why not storing them ordered?

0xDiscotech replied:
Will check how to do it

drgorillamd commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,344 @@
+  ) internal view returns (uint256 _amountOut) {
+    if (_wrapperName == INCH) {
+      _amountOut = OracleHelper.get1InchAmountOut(_wrapper, _tokenOne, _tokenTwo, _amountIn);
+      return _amountOut;
```
save some bytes when deploying by avoiding redundancy:)

0xng commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,344 @@
+  }
+
+  ///@inheritdoc IOracle
+  function addWrapperForPair(
```
Do you consider that the same way you can add a wrapper, you should be able to remove it? If I understand correctly, there's no way to stop supporting a given pair, you can only overwrite it with another wrapper. Mostly a question, not a feature request

0xDiscotech replied:
You're right, there is no way of removing a wrapper. I considered it but it wasn't requested in the description, so I moved ahead

drgorillamd commented on solidity/contracts/utils/OracleHelper.sol:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  // Get the curve indexes on the pool for the given tokens
+  function getCurveIdxs(
```
nit> with the other curve related function?

drgorillamd replied:
`ICurveRegistry(registry).get_coin_indices(pool, tokenIn, tokenOut);`

0xDiscotech replied:
Perfect, didn't know it exists

0xng commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,344 @@
+    }
+
+    // Ensure that the wrapper supports the given pair by calling the oracle
+    _getAmountByWrapperName(_wrapperName, _wrapper, _tokenOne, _tokenTwo, ONE_ETH);
```
This, as well as the dispatcher you built in this internal is clever and clean. Very nice. However, it seems to be lacking checks either here, in the internal, or in the library. Remember when working with external code to read what the different functions can return in odd cases, some may return a value instead of reverting when they don't support a pair, and that can lead to a bug in this contract

0xDiscotech replied:
Perfect, I will add a revert if there is not marching wrapper inside this function

drgorillamd commented on solidity/contracts/utils/OracleHelper.sol:
```diff
@@ -0,0 +1,106 @@
+   */
+  function getChainlinkAmountOut(address _wrapper, uint256 _amountIn) internal view returns (uint256 _amountOut) {
+    IChainlinkOracle _chainlinkWrapper = IChainlinkOracle(_wrapper);
+    (, int256 _amount,,,) = _chainlinkWrapper.latestRoundData();
```
this needs further validation (stale price etc)

drgorillamd commented on solidity/contracts/utils/OracleHelper.sol:
```diff
@@ -0,0 +1,106 @@
+  function getChainlinkAmountOut(address _wrapper, uint256 _amountIn) internal view returns (uint256 _amountOut) {
+    IChainlinkOracle _chainlinkWrapper = IChainlinkOracle(_wrapper);
+    (, int256 _amount,,,) = _chainlinkWrapper.latestRoundData();
+    _amountOut = _amountIn * uint256(_amount);
```
did you know the _amount returned by chainlink can be negative (hence the int256)?

0xDiscotech replied:
So in that case, returned should be 0, right?

0xng commented on solidity/contracts/utils/OracleHelper.sol:
```diff
@@ -0,0 +1,106 @@
+library OracleHelper {
+  using OracleLibrary for uint256;
+
+  function get1InchAmountOut(
```
For all these price-fetching functions, I would ask you to check these things:
- What happens if the pair I'm fetching is one composed of tokens with different numbers of decimals?
- Are they properly sanitized? Can you add any checks that may protect the code from returning a wrong, or stale price?
- Do you need to multiply the `_amountOut` by `_amountIn` other than for chainlink?

0xDiscotech replied:
* Fixed the case of different decimals, now it converts the amount in to wei and it returns always the response in wei

0xng commented on solidity/contracts/Oracle.sol:
```diff
@@ -0,0 +1,344 @@
+    bytes32 _tokensPairHash = keccak256(abi.encodePacked(_tokenOne, _tokenTwo));
+    _pairInfo = pairInfo[_tokensPairHash];
+
+    // If the pair is not found, try to get the pair with the tokens inverse order
```
Maybe you can benefit from token sorting like Uniswap does here

drgorillamd commented on solidity/contracts/utils/OracleHelper.sol:
```diff
@@ -0,0 +1,106 @@
+
+    int128 _tokenOneCurveIdx;
+    int128 _tokenTwoCurveIdx;
+    for (uint256 _i = 0; _i < 3; _i++) {
```
ICurveRegistry(registry).get_coin_indices(pool, tokenIn, tokenOut);

0xng reviewed (commented):
Very clean! Loved some of the ideas and implementations. Left some comments, but nothing too major.

One thing that perhaps is worth mentioning is that the approach you went for performs a ton of cold `SSTORES` every time `addWrapperForPair` is called. That's expensive in terms of gas, as each cold `SSTORE` is around 20k gas.

I won't ask you to implement it or anything, but can you theoretically come up with a different way of handling the code that would've allowed you to only have one or two storage variables?

drgorillamd commented on solidity/test/unit/Oracle.t.sol:
```diff
@@ -0,0 +1,500 @@
+}
+
+abstract contract Base is DSTestPlus {
+  uint256 public constant MAX_UINT = 2 ** 256 - 1;
```
type(uint256).max

0xDiscotech commented:
Added the suggested improvements and fixes suggested by @drgorillamd and @0xng:

* Sanitized oracles response
* Added assertions over internal functions and libraries
* Normalized amount in to wei in case the token has different decimals than 18
* Used curve function for getting coins
* Sorted addresses like uniswap does for getting the pair hash
* Added reverts with custom errors on internal functions too

0xng commented on solidity/contracts/utils/OracleHelper.sol:
```diff
@@ -37,24 +36,20 @@ library OracleHelper {
-      if (_curveWrapper.coins(_i) == _tokenTwo) {
-        _tokenTwoCurveIdx = int128(int256(_i));
-      }
+    (int128 _idxOne, int128 _idxTwo, bool _underlyingCoinOrder) =
```
does this return the underlyingCoinOrder or whether the tokens are actually underlying tokens in the pool?

0xDiscotech replied:
It returns the underlying order, I had the same question at first but after testing I realized that is the underlying order. But maybe I am wrong, in that case, just let me know 😄

0xng commented on solidity/contracts/utils/OracleHelper.sol:
```diff
@@ -37,24 +36,20 @@ library OracleHelper {
+    (int128 _idxOne, int128 _idxTwo, bool _underlyingCoinOrder) =
+      ICurveRegistry(_curveRegistry).get_coin_indices(_wrapper, _tokenOne, _tokenTwo);
+    if (_underlyingCoinOrder) {
+      _amountOut = _curveWrapper.get_dy(_idxOne, _idxTwo, _amountIn);
```
Does this revert if _idxOne and _idxTwo are not underlying tokens of the same pool, or does it return a weird value?

0xDiscotech replied:
I'm not sure tbh. But in case it returns a weird value, it will fail when asking for the price

0xng commented on solidity/contracts/Oracle.sol:
```diff
@@ -85,7 +85,10 @@ contract Oracle is Ownable, IOracle {
     // Ensure that the wrapper supports the given pair by calling the oracle
-    _getAmountByWrapperName(_wrapperName, _wrapper, _tokenOne, _tokenTwo, 1 ether);
+    uint256 _amountOut = _getAmountByWrapperName(_wrapperName, _wrapper, _tokenOne, _tokenTwo, 1 ether);
+    if (_amountOut == 0) {
```
You already have this check on `_getAmountByWrapperName`
